### PR TITLE
feat: debounced validation

### PR DESCRIPTION
**Pull request by pidkopajo**
Adds support for debounced validation.

### CODE REVIEW

File: src/core/types/form-validator.ts [hunk 1]
```diff
@@ -75,6 +77,7 @@ export type ValidateConfig<T, Err, Dependencies extends any[]> = {
   field: ValidateField<T, Err>;
   triggers?: ValidationTrigger[];
   dependencies?: readonly [...FieldDescTuple<Dependencies>];
+  debounce?: number;
   rules: (
```
Comment (mixvar) [inline src/core/types/form-validator.ts:81]: Lets add missing JSDocs for all properties here - will be quite helpful.
Should explain:
 - difference when passing `Schema.array` vs `Schema.array.nth`
 - effect of specifying triggers (and default - all triggers)
 - effect of specifying dependencies (available in rules cb; extra validation runs when dep changes)
 - effect of specifying debounce (mention that it affects all downstream rules for the field)

Comment (pidkopajo) [reply src/core/types/form-validator.ts:81]: wdyt of adding it to https://github.com/VirtusLab/formts/issues/52 ?

File: src/core/builders/create-form-validator.spec.ts [hunk 2]
```diff
@@ -1095,3 +1096,219 @@ describe("createFormValidator", () => {
...
+  it("should debounce when single-rule debounced validator called >1 times", async () => {
+    const stringRequiredValidator = jest.fn((x: string) =>
+      x ? null : "REQUIRED"
+    );
+
+    const { validate } = createFormValidatorImpl(Schema, validate => [
+      validate({
+        field: Schema.string,
+        rules: () => [stringRequiredValidator],
+        debounce: 1000,
```
Comment (mixvar) [inline src/core/builders/create-form-validator.spec.ts:1121]: the spec works nicely but using real debounce timers means it runs much longer than other specs
Screenshot 2021-02-22 at 00 32 11

maybe try https://jestjs.io/docs/en/timer-mocks (didn't really work last time I tried it) or at least decrease debounce values

File: src/core/builders/create-form-validator.ts [hunk 3]
```diff
@@ -308,3 +319,44 @@ type Timestamp = number;
...
+          debouncedValidations[id]?.();
+
+          const timeout = setTimeout(() => {
+            delete debouncedValidations[id];
+            resolve();
+          }, v.debounce);
+
+          debouncedValidations[id] = () => {
+            clearTimeout(timeout);
+            reject("cancel");
```
Comment (mixvar) [inline src/core/builders/create-form-validator.ts:346]: I'd extract this to const sth like `const CANCEL_ERR = "__CANCEL__"`

File: src/core/builders/create-form-validator.ts [hunk 4]
```diff
@@ -308,3 +319,44 @@ type Timestamp = number;
...
+  type Cancel = () => void;
+  type DebouncedValidationsDict = Record<FieldValidationKey, Cancel>;
+
+  export const createDebounceStepHandler = () => {
+    const debouncedValidations: DebouncedValidationsDict = {};
+
+    return (
+      v: FieldValidator<unknown, unknown, unknown[]>,
+      field: FieldDescriptor<unknown, unknown>
+    ) => {
```
Comment (mixvar) [inline src/core/builders/create-form-validator.ts:333]: maybe implicit return?

File: src/core/builders/create-form-validator.spec.ts (hunk 2 above)
Comment (pidkopajo) [reply src/core/builders/create-form-validator.spec.ts]: decreased debounced values
```
Test Suites: 28 passed, 28 total
Tests:       496 passed, 496 total
Snapshots:   0 total
Time:        2.558 s, estimated 3 s
```

**Review by mixvar — APPROVED** [review]

**Pull request merged**